File: src/content/docs/topics/cs/coding-problems/graphs/323-number-of-connected-components-in-an-undirected-graph-approach2.go
package main

import "fmt"

func countComponents(n int, edges [][]int) int {
	parent := make([]int, n)
	for i := range parent {
		parent[i] = i
	}
	count := n

	var find func(x int) int
	find = func(x int) int {
		for parent[x] != x {
			parent[x] = parent[parent[x]]
			x = parent[x]
		}
		return x
	}

	for _, e := range edges {
		ru, rv := find(e[0]), find(e[1])
		if ru != rv {
			parent[ru] = rv
			count--
		}
	}
	return count
}

func assert(condition bool, msgs ...string) {
	if !condition {
		msg := "assertion failed"
		if len(msgs) > 0 {
			msg = msgs[0]
		}
		panic(msg)
	}
}

func runTests() {
	assert(countComponents(5, [][]int{{0, 1}, {1, 2}, {3, 4}}) == 2)
	assert(countComponents(5, [][]int{{0, 1}, {1, 2}, {2, 3}, {3, 4}}) == 1)
	assert(countComponents(4, [][]int{}) == 4)
	assert(countComponents(1, [][]int{}) == 1)
	assert(countComponents(3, [][]int{{0, 1}, {1, 2}, {0, 2}}) == 1)
	fmt.Println("all tests pass")
}

func main() { runTests() }

File: src/content/docs/topics/cs/coding-problems/graphs/323-number-of-connected-components-in-an-undirected-graph-approach2.py
def count_components(n: int, edges: list[list[int]]) -> int:
    parent = list(range(n))
    count = n

    def find(x):
        while parent[x] != x:
            parent[x] = parent[parent[x]]
            x = parent[x]
        return x

    for u, v in edges:
        ru, rv = find(u), find(v)
        if ru != rv:
            parent[ru] = rv
            count -= 1
    return count

assert count_components(5, [[0,1],[1,2],[3,4]]) == 2
assert count_components(5, [[0,1],[1,2],[2,3],[3,4]]) == 1
assert count_components(4, []) == 4
assert count_components(1, []) == 1
assert count_components(3, [[0,1],[1,2],[0,2]]) == 1
print("all tests pass")
